service/history/replication/poller_manager.go

```diff
@@ -0,0 +1,83 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package replication
+
+import (
+	"fmt"
+
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/service/history/shard"
+)
+
+type (
+	pollerManagerImpl struct {
+		shard           shard.Context
+		clusterMetadata cluster.Metadata
+	}
+)
+
+func newPollerManager(
+	shard shard.Context,
+	clusterMetadata cluster.Metadata,
+) *pollerManagerImpl {
+	return &pollerManagerImpl{
+		shard:           shard,
+		clusterMetadata: clusterMetadata,
+	}
+}
+
+func (p pollerManagerImpl) getPollingShardIDs(remoteClusterName string) []int32 {
+	currentCluster := p.clusterMetadata.GetCurrentClusterName()
+	allClusters := p.clusterMetadata.GetAllClusterInfo()
+	currentClusterInfo, ok := allClusters[currentCluster]
+	if !ok {
+		panic("Cannot get current cluster info from cluster metadata cache")
+	}
+	remoteClusterInfo, ok := allClusters[remoteClusterName]
+	if !ok {
+		panic(fmt.Sprintf("Cannot get remote cluster %s info from cluster metadata cache", remoteClusterName))
+	}
+	return generatePollingShardIDs(p.shard.GetShardID(), currentClusterInfo.ShardCount, remoteClusterInfo.ShardCount)
+}
+
+func generatePollingShardIDs(localShardId int32, localShardCount int32, remoteShardCount int32) []int32 {
+	var pollingShards []int32
+	if remoteShardCount <= localShardCount {
+		if localShardId <= remoteShardCount {
+			pollingShards = append(pollingShards, localShardId)
```

Comment:
Just want to confirm my understanding. In this case, when applying the polled replication task, local cluster will re-calculate the shard ID and forward the task to the history host owning the re-calculated shard?

Reply:
Yes.